rules/do-you-have-separate-development-testing-and-production-environments/rule.md

```diff
@@ -1,80 +1,83 @@
 ---
 type: rule
-archivedreason: 
-title: Do you have separate development, testing and production environments?
-guid: 616246d0-1675-4c1c-b4b0-d4352fe818e1
+title: Do you have separate development, testing, and production environments?
 uri: do-you-have-separate-development-testing-and-production-environments
-created: 2009-03-10T07:02:13.0000000Z
 authors:
-- title: Adam Cogan
-  url: https://ssw.com.au/people/adam-cogan
-related: 
+  - title: Adam Cogan
+    url: https://ssw.com.au/people/adam-cogan
+  - title: Daniel Mackay
+    url: https://ssw.com.au/people/daniel-mackay
+related:
   - do-you-identify-development-test-and-production-crm-web-servers-by-colors
   - azure-resources-creating
   - apply-tags-to-your-azure-resource-groups
   - do-you-identify-development-test-and-production-crm-web-servers-by-colors
   - do-you-always-rename-staging-url-on-azure
 redirects: []
-
+created: 2009-03-10T07:02:13.000Z
+archivedreason: null
+guid: 616246d0-1675-4c1c-b4b0-d4352fe818e1
 ---
+It is important to separate production and non-production environments. Some companies skip the non-production environments because it can be a hassle to copy new files, register DLLs and deploy backend changes. This will usually result in higher support costs and unhappy users due to simple bugs that could have been found in testing.  
 
-It is important to maintain three separate environments for development, testing and production. Some companies skip the testing server because it can be a hassle to copy new files, register DLLs and deploy backend changes. This will usually result in higher support costs and unhappy users due to simple bugs that could have being found in testing.  
 <!--endintro-->
 
-::: greybox
-The old solution is to use build scripts (.bat and .vbs files) to automatically create a setup package that can be used to deploy to testing and production environments. For backend changes, you can either include the change scripts with the setup package (if it's a localised database), or run those scripts as part of your deployment process.
+## What is each environment for?
 
-Read more about setup packages at [SSW's Wise Standard for Products.](http://www.ssw.com.au/ssw/Standards/wisesetup/WiseStandards.aspx)
-:::
+* **Production**: Real data being used by real customers
+* **Staging**: 'Production-like' data.  Used for testing and verification before deploying to Production.  Should have 'production-like' access (i.e. avoid giving devs admin rights).  Should generally have the same specs as production (especially during performance testing).  However, this is not always the case due to cost implications
+* **Test**: Internal environment used by the development team.  Hosted in the same place as the Staging & Production environments.  Access usually not provided to the client
```

Comment:
```suggestion
* **Test**: A place to verify code changes. Typically simpler or under-specified version of Staging or Production environments aiding in the early identification and troubleshooting of issues.
```
> Internal environment used by the development team.
> Access usually not provided to the client
I'm not sure about these comments, why does it need to be so prescriptive

Reply:
The client still has access to the dev/test environments (they're paying for it). We give them the understanding that this environment is for the dev team's test purposes and that there is no guarantee that this is a stable environment.

Reply:
Have removed the comment about client access to 'Test' as per the suggestion from @wicksipedia.